### DISCUSSION TITLE
Integrating combobulate with meow

### DISCUSSION THREAD

**Discussion by jgarte** [meow-edit/meow#448] [Q&A] [answered]
Hi,

This [article](https://www.masteringemacs.org/article/combobulate-structured-movement-editing-treesitter) on combobulate mentions having support for [mutliple-cursors ](https://github.com/magnars/multiple-cursors.el )package.

How are meow users thinking about integrating multiple cursors with combobulate?

**Comment by DogLooksGood** [comment, marked as answer]
There's always an integration possibility. But I won't say multiple-cursors is good for meow, since we embrace kmacros.

As for tree-sitters and structure editing, we can either build our own meow state based on tree-sitter itself, or leave users to create their own dedicated state for modes like combobulate. Meow has no parentheses and structure editing features builtin, there are so many options available.

**Reply by jgarte** [reply]
I should have phrased my question as the following

> How are meow users thinking about integrating their own kmacro multiple cursors with combobulate?

> But I won't say multiple-cursors is good for meow, since we embrace kmacros.

I wasn't thinking for meow to integrate with multiple-cursors. Sorry for the confusion on that.

**Reply by jgarte** [reply]
> there are so many options available.

Yup, I guess the issue is deciding on one of the options and making it convenient for users to use tree-sitter and possibly combobulate with meow.

**Reply by DogLooksGood** [reply]
I should give a try on combobulate.

**Reply by jgarte** [reply]
Yeah, me too...

**Comment by jgarte** [comment]
Some (non)inspiration:

https://github.com/meain/evil-textobj-tree-sitter

**Comment by gs-101** [comment]
If anyone is interested on this, I think I got a nice set-up for structured navigation, using `sexp` and `list` for general mode, and replacing them with combobulate-specific binds when it is activated.

```emacs-lisp
(use-package meow-helpers
  :config
  (meow-normal-define-key
   '("H" . backward-sexp)
   '("J" . down-list)
   '("K" . backward-up-list)
   '("L" . forward-sexp)
   '("S" . kill-sexp)
   '("t" . meow-transpose-sexp)
   ))
```

```emacs-lisp
(use-package combobulate
  :bind
  (
   :map combobulate-key-map
   ([remap backward-up-list] . combobulate-navigate-up)
   ([remap down-list] . combobulate-navigate-down)
   ([remap forward-sexp] . combobulate-navigate-next)
   ([remap backward-sexp] . combobulate-navigate-previous)
   ([remap transpose-sexp] . combobulate-transpose-sexps)
   ([remap kill-sexp] . combobulate-kill-node-dwim)
   ))
```

EDIT: My source of inspiration was [this](https://karthinks.com/software/a-consistent-structural-editing-interface/) blog post by [karthink](https://github.com/karthink).

**Reply by gs-101** [reply]
I just discovered this, treesit includes `treesit-forward-sexp`, but without a key-map I don't know how to effectively remap it.

**Reply by DogLooksGood** [reply]
Hey, what do you want to remap? To remap by iterating a keymap?

**Reply by gs-101** [reply]
I wanted to remap `forward-sexp` to `treesit-forward-sexp` in treesit modes only, but couldn't find a `treesit-key-map`. This is what I did with combobulate in the example above, `forward-sexp` is remaped to a combobulate function only in combobulate buffers.

EDIT: Maybe it is already included in a newer version of Emacs, It's been about a month since I compiled it from source.

EDIT 2: I found this in `treesit.el.gz`:

```emacs-lisp
  (when (treesit-thing-defined-p 'sexp nil)
    (setq-local forward-sexp-function #'treesit-forward-sexp)
    (setq-local transpose-sexps-function #'treesit-transpose-sexps))
```

So I guess this remapping isn't necessary, as treesit itself replaces `forward-sexp`.

**Reply by gs-101** [reply]
Further development on this topic, I'll be using something like this now:

```emacs-lisp

(use-package meow-command
  :config
  (defun gs-101/meow-super-prev ()
    "Runs different commands based on the current major or minor mode.

- `combobulate-mode' :: `combobulate-navigate-up'
- `markdown-mode' :: `markdown-previous-visible-heading'
- `org-mode' :: `org-previous-visible-heading'
- `prog-mode' :: `backward-up-list'
- Other :: `meow-prev-expand'"
    (interactive)
    (cond
     ((seq-some (lambda (mode) (string-match-p "combobulate" (symbol-name mode))) local-minor-modes) (combobulate-navigate-up))
     ((string-match-p "markdown-mode\\'" (symbol-name major-mode)) (markdown-previous-visible-heading))
     ((string-match-p "org-mode\\'" (symbol-name major-mode)) (org-previous-visible-heading 1))
     ((derived-mode-p 'prog-mode) (backward-up-list))
     (t (meow-prev-expand)))
    )
  )
```

Keeps everything right inside Meow.
